Feature - add message body handlers for message processing (#124)
arcus-azure/arcus.messaging: docs/preview/features/message-pumps/customization.md

> 

Reviewer: Can you update this TOC with the subheaders please?

> ## Control custom deserialization

Reviewer: I would move this after "Filter messages based on message context" & "Fallback message handling" because it will be less used, unless you want to keep this for alphabetically sorting (which is ok)

> ### Altering custom message pump

Reviewer: ```suggestion
### Extending the message pump
```

}
```

> ### Delegating deserialization

Reviewer: ```suggestion
### Bring your own deserialization
```

> You can also choose to delegate the message deserialization to different deserializers. 

Reviewer: ```suggestion
You can also choose to extend the built-in message deserialization with additional deserializer to meet your needs. 
```

> This allows for more reuse, asynchronous, and is available without altering your message pump. 

Reviewer: What about this?
```suggestion
This allows you to easily deserialize into different message formats or reuse existing (de)serialization capabilities that you already have without altering your message pump. 
```

You start by implemeting an `IMessageBodyHandler`. The following example shows how an expected type can be transformed to something else. 
The result type (in this case `OrderBatch`) will be then be used to check if there is an `IMessageHandler` registered with that message type.

```csharp

> public class OrderBatchMessageBodyHandler : IMessageBodyHandler

Reviewer: Wouldn't it make more sense to emphasize that it's only for serialization?
```suggestion
public class OrderBatchMessageBodySerializer : IMessageBodySerializer
```

{
    public async Task<MessageResult> DeserializeMessageAsync(string messageBody)
    {

>         var orders = JsonConvert.DeserializeObject<Order[]>(messageBody);

Reviewer: Maybe it's more interesting to use an XML example here to more show its use?

        return MessageResult.Success(new OrderBatch(orders));
    }
}
```

The registration of these message body handlers can be done just as easily as an `IMessageHandler`:

```csharp
public void ConfigureServices(IServiceCollection services)

> {

Reviewer: Shouldn't the registration be linked to the message pump that is being registered? Either generally or per message handler because otherwise we will do a lot of deserialization to match things, no?

```csharp
services.AddServiceBusQueueMessagePump("ARCUS_SERVICEBUS_ORDERS_CONNECTIONSTRING")
            .WithServiceBusMessageHandler<OrdersMessageHandler, Order, OrderBatchMessageBodySerializer>();
```

I'm just thinking out loud.

Author: We could do that, yeah, we should then register the message handler differently but it's possible.

Author: Should I move to this approach?

Author: Moved to this approach, resolving.

You start by implemeting an `IMessageBodySerializer`. The following example shows how an expected type can be transformed to something else. 
The result type (in this case `OrderBatch`) will be then be used to check if there is an `IMessageHandler` registered with that message type.

```csharp
public class OrderBatchMessageBodySerializer : IMessageBodySerializer
{
    public async Task<MessageResult> DeserializeMessageAsync(string messageBody)
    {
        var serializer = new XmlSerializer(typeof(Order[]));
        using (var contents = new MemoryStream(Encoding.UTF8.GetBytes(messageBody)))
        {
            var orders = (Order[]) serializer.Deserialize(contents);
            return MessageResult.Success(new OrderBatch(orders));
        }
    }
}
```

The registration of these message body handlers can be done just as easily as an `IMessageSerializer`:

```csharp
public void ConfigureServices(IServiceCollection services)
{
    // Register the message body serializer in the dependency container where the dependent services will be injected.

>     services.WithMessageHandler<OrderBatchMessageHandler>(..., messageBodySerializer: new OrderBatchMessageBodySerializer());

Reviewer: Since it's just a new here, why not do this?
```suggestion
    services.WithMessageHandler<OrderBatchMessageHandler, OrderBatchMessageBodySerializer>(...);
```

Author: Because we don't have such an extension 😄 . We could, but it seemed at the time an unnecessary generic. I could add it, but it would mean we create arround 24 extra extensions with all the combinations. (plus tests of course).

Reviewer: It feels more natural to me, but not worth the effort

Author: We could also change the existing ones so that there's a generic overload that creates it directly, and a implementation factory where you get the `IServiceProvider` to create your serializer.

Reviewer: Can you give an example please?

Author: Now, we have these capabilities:
- `.WithMessageHandler(..., messageBodySerializer: new OrderSerializer());`
- `.WithmessageHandler(..., messageBodySerializerImplementationFactory: serviceProvider => new OrderSerializer());`

We could change it throughout all combinations with the generic instead:
- `.WithMessageHandler<..., OrderSerializer>(...);`
- `.WithmessageHandler(..., messageBodySerializerImplementationFactory: serviceProvider => new OrderSerializer());`

Reviewer: I'm confused because isn't that what I suggested 🤔

Author: Well, I said we could _add_ also a new overload so that you could both send in the instance as the type and that would lead to 24+ extra extensions, or we could change the existing ones. Not sure what you suggested.

Author: Hmm, also, we're not gonna be able to do this approach for every overload bc it will result in some ambigous signatures (same amount of params, only with different generics, still fail).

Reviewer: Let's just resolve this for now :)
